data-prepper-plugins/in-memory-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/inmemory/InMemoryPartitionAccessor.java

```diff
@@ -0,0 +1,94 @@
+        this.partitionLookup = new HashMap<>();
+    }
+
+    public Optional<SourcePartitionStoreItem> getItem(final String partitionKey) {
+
+        final InMemorySourcePartitionStoreItem item = partitionLookup.get(partitionKey);
+
+        if (Objects.isNull(item)) {
+            return Optional.empty();
+        }
+
+        return Optional.of(item);
+    }
+
+    public void queuePartition(final InMemorySourcePartitionStoreItem inMemorySourcePartitionStoreItem) {
+        queuedPartitions.add(inMemorySourcePartitionStoreItem.getSourcePartitionKey());
+        partitionLookup.put(inMemorySourcePartitionStoreItem.getSourcePartitionKey(), inMemorySourcePartitionStoreItem);
+    }
+
+    public Optional<SourcePartitionStoreItem> getNextItem() {
+
+        final Set<String> seenPartitionKeys = new HashSet<>();
+        while (!queuedPartitions.isEmpty()) {
+            final String nextPartitionKey = queuedPartitions.peek();
+
+            // Keep from having infinite loop (or infinite until a reopenAt timestamp is met) when only CLOSED partitions are left
+            if (seenPartitionKeys.contains(nextPartitionKey)) {
+                break;
+            }
+            seenPartitionKeys.add(nextPartitionKey);
+
+            final InMemorySourcePartitionStoreItem nextItem = partitionLookup.get(nextPartitionKey);
+            if (Objects.isNull(nextItem)) {
+                queuedPartitions.remove();
+            } else if (SourcePartitionStatus.CLOSED.equals(nextItem.getSourcePartitionStatus())
+                            && Objects.nonNull(nextItem.getReOpenAt())
+                            && nextItem.getReOpenAt().isAfter(Instant.now())) {
+                // Put to the back of the queue when partition is closed and has not reached reopenAt instant
+                queuedPartitions.remove();
+                queuedPartitions.add(nextPartitionKey);
+            } else {
+                queuedPartitions.remove();
+                return Optional.of(nextItem);
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    public void updateItem(final InMemorySourcePartitionStoreItem item) {
+        final InMemorySourcePartitionStoreItem itemToUpdate = partitionLookup.get(item.getSourcePartitionKey());
+
+        // If marking as unassigned or closed means the request is to process the next partition, so will requeue this one
+        if (SourcePartitionStatus.UNASSIGNED.equals(item.getSourcePartitionStatus())
+                || SourcePartitionStatus.CLOSED.equals(item.getSourcePartitionStatus())) {
+            queuedPartitions.add(item.getSourcePartitionKey());
+        }
+
+        // The partition lookup map grows infinitely as more and more partitions are created
+        // Should we remove COMPLETED partition items from memory and keep track of them with a Set of partitionKeys?
+        // Will still work fine regardless if there aren't thousands of partitions, which seems ok since it is advised for non-production use
```

Comment:
Thanks @graytaylor0. I know we plan on only recommending this for non-production uses but I am concerned some users will still use this in production. I understand that we need to keep track of all partitions and that removing any will cause them to rescan. I am wondering if there is some strategies we can use here to help out users migrate away from this:

1. Add a warning message on start up to raise that this is not meant for production workloadsCap the in memory usage.
2. Create a clean up policy for partitions older than x number of days
3. Create a clean up policy for partitions which no longer exist in the source? This is probably difficult and not worth it for all the different sources
4. Use a file cache for old data...

I think 4 is the easiest for now. We can always improve on this later.

Reply:
I think a file cache could work.

Conceptually however, I think we should consider how source coordination works for continually growing sources. I tend to think that we should expect some sort of "ordering" from the source. This could be provided by pipeline authors in some cases (e.g. `SORT BY`) or could be part of the source itself (e.g. `ListObjects` in S3 returns in a certain order).